File: src/main/java/scraping/models/Semester.java
package scraping.models;

public enum Semester {
  January(2),
  Spring(4),
  Summer(6),
  Fall(8);

  private Integer id;

  Semester(Integer id) { this.id = id; }

  public static Semester fromCode(String code) {
    Semester sem;
    switch (code) {
    case "ja":
      sem = January;
      break;
    case "fa":
      sem = Fall;
      break;
    case "sp":
      sem = Spring;
      break;
    case "su":
      sem = Summer;
      break;
    default:
      sem = valueOf(code);
    }
    return sem;
  }

  public static Semester fromInt(Integer id) {
    Semester sem;
    switch (id) {
    case 2:
      sem = January;
      break;
    case 4:
      sem = Spring;
      break;
    case 6:
      sem = Summer;
      break;
    case 8:
      sem = Fall;
      break;
    default:
      throw new IllegalArgumentException(
          "ID can only be one of {2,4,6,8} (got " + id + ")");
    }
    return sem;
  }

  public Integer toInt() { return this.id; }
}

File: backend/src/models/Semester.kt
package schedge.models

enum class Semester {
  January,
  Spring,
  Fall,
  Summer;

  companion object {
    fun from_int(id: Int): Semester {
      return when (id) {
        2 -> Semester.January
        4 -> Semester.Spring
        6 -> Semester.Summer
        8 -> Semester.Fall
        else -> throw IllegalArgumentException("Used invalid id")
      }
    }
  }

  fun as_int(): Int {
    return when (this) {
      Semester.January-> 2
      Semester.Spring -> 4
      Semester.Summer -> 6
      Semester.Fall -> 8
    }
  }
}


/**
 * A term, like Summer 19. year is a positive number representing the year.
 * For example, 2019 represents the year 2019.
 */
class Term(val sem: Semester, val year: Int) {

  companion object {
    fun from_id(id: Int): Term {
      val year = id / 10
      return Term(Semester.from_int(id - year), year)
    }
  }

  init {
    if (year < 0)
      throw IllegalArgumentException("Can't create a Term with negative year")
  }

  val id
    get(): Int {
      return (year - 1900) * 10 + sem.as_int()
    }

  override fun toString(): String {
    return "Term(${sem}, ${year})"
  }
}
